### PR TITLE
Add Support for Admin Network Policy

### LINKED ISSUE #3158: Implement Admin Cluster Network Policy
Tracks trying to implement https://raw.githubusercontent.com/kubernetes-sigs/network-policy-api/master/config/crd/bases/policy.networking.k8s.io_adminnetworkpolicies.yaml in OVNK

### PR DESCRIPTION
**Pull request by tssurya**
**EDIT**: It was decided in the team meeting to change all our controllers into being like the `service` controller where every object change is re-queued back to the main queue and processed in a single place. ANP will also follow this pattern. I am re-working this as a new PR. Hence this PR is stale and will be superseded by https://github.com/ovn-org/ovn-kubernetes/pull/3659

Not ready for review, development ongoing.
https://docs.google.com/presentation/d/1Y-PrygN5vfDet0K3l89o83kJ3ryial3Dq2ZE3BDVlzM/edit#slide=id.p captures the implementation details in OVN-Kubernetes

**Tasks**
- [x] Vendor in network-policy-api
- [x] Add a feature enable flag
- [x] Add Controller for ANP
- [x] Implement add/delete/update ANP
- [x] Implement add/delete/update BANP
- [x] Implement selectors for peers
- [x] Implement namespace label updates for ANP
- [x] Implement pod label updates for ANP
- [x] Implement namespace deletion for ANP
- [x] Implement pod deletion and take care of completed pods logic for ANP
- [x] Implement namespace label updates for BANP
- [x] Implement pod label updates for BANP
- [x] Implement namespace deletion for BANP
- [x] Implement pod deletion and take care of completed pods logic for BANP
- [ ] Implement sameLabels/notSameLabels for peers - This will be done only in next iteration (betav1 support)
- [ ] Implement Named Ports for Conformance - depends on time, not a big deal to get in in this iteration
- [x] Ensure external_ids are set correctly for ACLs, address-sets, portgroups
- [x] Implement PASS action
- [x] Implement Status field in ANP/BANP
- [x] Write repair function
- [ ] Fix all TODO's in code

**Tests**
- [x] Add unit testing for ANP
- [ ] Add cache testing for ANP
- [x] Add unit testing for BANP
- [ ] Add cache testing for BANP
- [ ] Add status testing for ANP/BANP
- [ ] Add NamedPort testing for ANP/BANP
- [ ] Add e2e testing for ANP - https://github.com/kubernetes-sigs/network-policy-api/pull/99 and https://github.com/kubernetes-sigs/network-policy-api/pull/112 and https://issues.redhat.com/browse/SDN-3975
- [ ] Add e2e testing for BANP - https://github.com/kubernetes-sigs/network-policy-api/pull/109 and https://github.com/kubernetes-sigs/network-policy-api/pull/112
- [x] Add e2e testing for interaction with NP - https://github.com/kubernetes-sigs/network-policy-api/pull/105
- [x] Overall testing - Bring all the upstream tests middlestream here into this PR

**Improvements**
- [ ] Perf improvement - use existing address-sets for namespaces without creating new ones

**- What this PR does and why is it needed**
This PR adds support for admin network policy in ovn-kubernetes. The API lives at https://github.com/kubernetes-sigs/network-policy-api and is managed by sig-network-policy-api working group. It is currently in v1alpva1 version. Hence until it graduates to beta, we probably will have only a tech preview version of this, not a full GA-ed version.

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

Closes #3158

### CODE REVIEW

**Comment by tssurya** [comment]
```
2023-03-28T10:53:07.1750963Z I0328 10:53:07.164221   20835 gateway_shared_intf.go:562] Adding service service1 in namespace namespace1
2023-03-28T10:53:07.1751378Z I0328 10:53:07.164266   20835 gateway_shared_intf.go:583] Rules already programmed for service1 in namespace namespace1
2023-03-28T10:53:07.1751639Z 
2023-03-28T10:53:07.1751838Z [91m[1m• Failure [0.119 seconds][0m
2023-03-28T10:53:07.1752065Z Node Operations
2023-03-28T10:53:07.1752538Z [90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_localnet_linux_test.go:219[0m
2023-03-28T10:53:07.1752852Z   on add
2023-03-28T10:53:07.1756119Z   [90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_localnet_linux_test.go:377[0m
2023-03-28T10:53:07.1756904Z     [91m[1minits iptables rules and openflows with LoadBalancer where AllocateLoadBalancerNodePorts=False, ETP=local, LGW mode [It][0m
2023-03-28T10:53:07.1757645Z     [90m/home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_localnet_linux_test.go:837[0m
2023-03-28T10:53:07.1757890Z 
2023-03-28T10:53:07.1758100Z     [91mExpected success, but got an error:
2023-03-28T10:53:07.1758475Z         <*errors.errorString | 0xc000e14230>: {
2023-03-28T10:53:07.1760212Z             s: "expected rule \"-p TCP -d 5.5.5.5 --dport 80 -j DNAT --to-destination 10.244.0.3:8080 -m statistic --mode random --probability 0.5000000000\" at pos 0 in chain nat/OVN-KUBE-ETP, got \"-p TCP -d 5.5.5.5 --dport 80 -j DNAT --to-destination 10.244.0.4:8080 -m statistic --mode random --probability 0.5000000000\"",
2023-03-28T10:53:07.1760752Z         }
2023-03-28T10:53:07.1762142Z         expected rule "-p TCP -d 5.5.5.5 --dport 80 -j DNAT --to-destination 10.244.0.3:8080 -m statistic --mode random --probability 0.5000000000" at pos 0 in chain nat/OVN-KUBE-ETP, got "-p TCP -d 5.5.5.5 --dport 80 -j DNAT --to-destination 10.244.0.4:8080 -m statistic --mode random --probability 0.5000000000"[0m
2023-03-28T10:53:07.1762563Z 
2023-03-28T10:53:07.1763018Z     /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/pkg/node/gateway_localnet_linux_test.go:957
2023-03-28T10:53:07.1763417Z [90m------------------------------[0m
2023-03-28T10:53:07.1763846Z [0mNode Operations[0m [90mon add[0m 
2023-03-28T10:53:07.1764300Z   [1minits iptables rules and openflows with LoadBalancer where ETP=cluster, LGW mode[0m
```
the pod's positions are changing or what?! ugh.

File: go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy_controller.go [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+		return false
+	}
+	defer c.anpQueue.Done(anpKey)
+
+	err := c.syncAdminNetworkPolicy(anpKey.(string))
+	if err == nil {
+		c.anpQueue.Forget(anpKey)
+		return true
+	}
+	klog.Infof("SURYA")
```
Comment (astoycos) [inline go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy_controller.go:279]:
```suggestion

```

File: go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go [hunk 2]
```diff
@@ -0,0 +1,664 @@
...
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing Admin Network Policy %s : %v", anpName, time.Since(startTime))
+	}()
+
+	anp, err := c.anpLister.Get(anpName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.Infof("SURYA %v", anp)
```
Comment (astoycos) [inline go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go:298]: nit comment

**Pull request closed without merging**